lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -235,4 +235,33 @@ sub save_results {
     path($self->job->result_dir, 'details-' . $self->name . '.json')->spurt(encode_json($results));
 }
 
+# Collect text output into JSON. save_results() is called too early for this.
+sub finalize_results {
+    my ($self) = @_;
+
+    my $dir = $self->job->result_dir();
+    return unless $dir;
+    my $file    = path($dir, "details-" . $self->name . ".json");
+    my $tmpfile = $file->sibling($file->basename . '.tmp');
+    return unless -e $file;
+    my $results;
+
+    eval {
+        $results = decode_json($file->slurp);
+        $results = {details => $results} if ref($results) eq 'ARRAY';
+
+        for my $step (@{$results->{details}}) {
+            next unless $step->{text};
+            my $txtfile = path($dir, $step->{text});
+            next unless -e $txtfile;
+            $step->{text_data} = $txtfile->slurp;
+        }
+
+        $tmpfile->spurt(encode_json($results));
+        rename $tmpfile->to_string, $file->to_string;
```

Comment:
Looks like you're missing exception handling. Probably best to just use the portable [move_to](https://metacpan.org/pod/Mojo::File#move_to) anyway.

Reply:
No. The two files are in the same directory so plain `rename` will work. `move_to` might create corrupted files which is specifically what I'm trying to avoid with this tempfile+rename trick.

Reply:
Why does `move_to` create a corrupted file? It will also just do a `rename` internally as far as i can see, and give you free exception handling.

Reply:
If `move_to` didn't do the most optimal operations possible for the case at hand it would be an upstream bug for me to fix.

Reply:
`move_to` [documentation](https://metacpan.org/pod/Mojo::File#move_to):

> Move file with File::Copy and return the destination as a Mojo::File object.

`File::Copy::move` [documentation](https://metacpan.org/pod/File::Copy#DESCRIPTION):

> If possible, move() will simply rename the file. Otherwise, it copies the file to the new location and deletes the original. **If an error occurs during this copy-and-delete process, you may be left with a (possibly partial) copy of the file under the destination name.**

Perl docs explicitly say that `rename` will be translated to the regular C syscall. That's exactly what I want. I have no reason to put several layers of additional abstraction between my code and the simple syscall that I want to execute. The only thing I get from that is the risk of unexpected bugs.

Reply:
In this case `File::Copy::move` will not perform a copy and delete however, so that does not apply. It's just a fallback for more portability. If you insist on using `rename` anyway you will have to add exception handling though.